libraries/mediaviewer/impl/src/main/kotlin/io/element/android/libraries/mediaviewer/impl/gallery/TimelineMediaItemsFactory.kt

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2023, 2024 New Vector Ltd.
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ * Please see LICENSE in the repository root for full details.
+ */
+
+package io.element.android.libraries.mediaviewer.impl.gallery
+
+import com.squareup.anvil.annotations.ContributesBinding
+import io.element.android.libraries.androidutils.diff.DiffCacheUpdater
+import io.element.android.libraries.androidutils.diff.MutableListDiffCache
+import io.element.android.libraries.core.coroutine.CoroutineDispatchers
+import io.element.android.libraries.di.AppScope
+import io.element.android.libraries.matrix.api.timeline.MatrixTimelineItem
+import io.element.android.libraries.mediaviewer.impl.gallery.diff.TimelineMediaItemsCacheInvalidator
+import io.element.android.services.toolbox.api.systemclock.SystemClock
+import kotlinx.collections.immutable.ImmutableList
+import kotlinx.collections.immutable.toPersistentList
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.withContext
+import timber.log.Timber
+import javax.inject.Inject
+
+interface TimelineMediaItemsFactory {
+    val timelineItems: Flow<ImmutableList<MediaItem>>
+
+    suspend fun replaceWith(timelineItems: List<MatrixTimelineItem>)
+    suspend fun onCanPaginate()
+}
+
+@ContributesBinding(AppScope::class)
+class DefaultTimelineMediaItemsFactory @Inject constructor(
+    private val dispatchers: CoroutineDispatchers,
+    private val virtualItemFactory: VirtualItemFactory,
+    private val eventItemFactory: EventItemFactory,
+    private val systemClock: SystemClock,
+) : TimelineMediaItemsFactory {
+    private val _timelineItems = MutableSharedFlow<ImmutableList<MediaItem>>(replay = 1)
+    private val lock = Mutex()
+    private val diffCache = MutableListDiffCache<MediaItem>()
+    private val diffCacheUpdater = DiffCacheUpdater<MatrixTimelineItem, MediaItem>(
+        diffCache = diffCache,
+        detectMoves = false,
+        cacheInvalidator = TimelineMediaItemsCacheInvalidator()
+    ) { old, new ->
+        if (old is MatrixTimelineItem.Event && new is MatrixTimelineItem.Event) {
+            old.uniqueId == new.uniqueId
+        } else {
+            false
+        }
+    }
+
+    override val timelineItems: Flow<ImmutableList<MediaItem>> = _timelineItems.distinctUntilChanged()
+
+    override suspend fun replaceWith(
+        timelineItems: List<MatrixTimelineItem>,
+    ) = withContext(dispatchers.computation) {
+        lock.withLock {
+            diffCacheUpdater.updateWith(timelineItems)
+            buildAndEmitTimelineItemStates(timelineItems)
+        }
+    }
+
+    /**
+     * Update the timestamp of the loading indicator, so that it may trigger a new pagination request.
+     */
+    override suspend fun onCanPaginate() {
```

Comment:
Why do we need this?

Reply:
It is used to `Update the timestamp of the loading indicator, so that it may trigger a new pagination request.` (see the comment in the impl of this fun)
The UI will trigger again a pagination request if the timeline does not load new event and the indicator is still displayed on the screen.

Reply:
Maybe this should be done in the Timeline implementation instead then, when the pagination request finishes, trigger a new build of items, even if there is no new ones

Reply:
[e5b3421](https://github.com/element-hq/element-x-android/pull/4010/commits/e5b34216e9c6ac83ddb808089912b6db8db865b6)